language/vm/src/binary_format.rs

```diff
@@ -0,0 +1,366 @@
+pub type IdentifierPool = Vec<Identifier>;
+// the pool of signatures, single pool where every signature is serialized as a single blob.
+// Argument signatures and return signatures are split in 2 blobs (see FunctionHandle)
+pub type Signature = Vec<Type>;
+pub type SignaturePool = Vec<Signature>;
+
+//
+// next 2 tables could be replaced by the constant pool (described at the bottom of this file)
+
+// The pool of byte array - Vector<u8> in Move
+pub type U8VectorPool = Vec<Vec<u8>>;
+// The address pool
+pub type AddressPool = Vec<AccountAddress>;
+
+//
+// Scripts and Modules
+//
+
+
+// Scripts and Modules have a common "section" expressed in Rust.
+// The Rust representation may vary... don't know...
+// The split is entirely an abstraction invention because at binary format we only have
+// tables. The order of tables can be anything.
+
+// Common section
+pub struct CompiledHeader {
+    // handles are references to definitions within and cross modules
+    pub module_handles: Vec<ModuleHandle>,
+    pub struct_handles: Vec<StructHandle>,
+    pub function_handles: Vec<FunctionHandle>,
+    pub field_handles: Vec<FieldHandle>,
+    // instantiations are references to a *generic* handle and and its instantiation, whether
+    // partial or complete
+    pub struct_instantiations: Vec<StructInstantiation>,
+    pub function_instantiations: Vec<FunctionInstantiation>,
+    pub field_instantiations: Vec<FieldInstantiation>,
+    // common blob tables
+    pub signatures: SignaturePool,
+    pub identifiers: IdentifierPool,
+    // constants
+    pub byte_array_pool: U8VectorPool,
+    pub address_pool: AddressPool,
+}
+
+// In `CompiledScript` we replace the FunctionDefinition for main(..) with broken up data
+// to set up the call frame and define/verify main signature and arguments
+pub struct CompiledScript {
+    pub header: CompiledHeader,
+
+    // Below should be enough information to materialize a
+    // main<Cup<T>, W>(arg1, ..., argn) {}
+    // with args of any type (including generic types) that the VM (verification
+    // and runtime) can use.
+    // There is a verification step here that is *purely* runtime.
+    // The client is passing the type arguments and you can imagine restrictions
+    // given our resource semantic.
+
+    // function arguments signature, main is void on return
+    pub arg_types: SignatureIndex,
+    // type parameter for the script
+    pub type_params: Vec<Kind>,
+    // code
+    pub main: CodeUnit,
+}
+
+// A CompiledModule is what it used to be but you will notice the removal of `FieldDefinition`s.
+// Fields are now serialized with the `StructDefinition` which changes verification around fields.
+// (see `StructDefinition` for details)
+pub struct CompiledModule {
+    pub header: CompiledHeader,
+
+    pub struct_defs: Vec<StructDefinition>,
+    pub function_defs: Vec<FunctionDefinition>,
+}
+
+//
+// Basics...
+//
+
+pub enum Kind {
+    All,
+    Resource,
+    Unrestricted,
```

Comment:
No strong feelings but the name `Unrestricted` has bothered me. 
To me, `Unrestricted` implies a "normal" type in a otherwise linear type system. This would mean that Move + Copy would become meaningless (where really everything is a copy). 
Since MoveLoc exists for these kinds, I might suggest something like `Copyable`, or `Affine` if you want to be pedantic  
Any thoughts @sblackshear ?

As a footnote:
In the source language, I use `Unrestricted` to refer to normal types that can be implicitly copied. `Affine` then refers to types that need to be explicitly copied (structs and vectors)

Reply:
I guess I like `Capyable`...

Reply:
Copyable sounds good to me (especially if that's what the source syntax says).